nixos/modules/virtualisation/amazon-init.nix

```diff
@@ -12,11 +12,26 @@ let
 
     userData=/etc/ec2-metadata/user-data
 
+    isScript() {
+        local fn="$1"
+        local fd
+        local magic
+        exec {fd}< "$fn"
+        read -r -n 2 -u "$fd" magic
+        exec {fd}<&-
+        if [[ "$magic" =~ \#! ]]; then return 0; else return 1; fi
+    }
+
     if [ -s "$userData" ]; then
-      # If the user-data looks like it could be a nix expression,
-      # copy it over. Also, look for a magic three-hash comment and set
-      # that as the channel.
-      if sed '/^\(#\|SSH_HOST_.*\)/d' < "$userData" | grep -q '\S'; then
+      if isScript "$userData"; then
+        # If the user-data is a script, run it.
+        echo "user data is a script; running"
+        chmod a+x "$userData"
+        exec "$userData"
```

Comment:
Any reason you can't do something like:

```
exec $SHELL "$userData"
```

BTW, is this `exec` intentional? It seems it will skip the rest of the script.

Reply:
Now looking at the rest of the module, it kinda makes more sense. But I don't know if I like this "magic" behavior. I would much prefer a explicit approach (like having an option with an enum that you could pass a indicating if userData is a bash script, user expression, etc).